File: problem72/Problem72.java
import java.util.*;
/**
 * Problem 72
 */

public class Problem72 {
	public static void main(String[] arg) {
		int upperBound = (int)Math.sqrt(1000000);
		boolean[] prime = IntegerArithmetic.sieveEratosthenes(upperBound);
		int nbPrimes = 0;
		for (int i = 0; i < prime.length; i++) {
			if (prime[i]) nbPrimes++;
		}
		int[] primes = new int[nbPrimes];
		int j = 0;
		for (int i = 0; i < prime.length; i++) {
			if (prime[i]) {
				primes[j] = i;
				j++;
			}
		}	
		/*int upperBound = 1000000;
		int k = (int)Math.sqrt(upperBound);
		boolean[] prime = IntegerArithmetic.sieveEratosthenes(k);*/
		long sum = 0;
		for (int d = 2; d <= 1000000; d++) {
			sum += (long)phi(d, primes);
		}
		System.out.println("The solution is " + sum);
	}
	
	/**
	* Returns phi n
	*/
	public static int phi(int n, int[] primes) {
		ArrayList<ArrayList<Integer>> primeDec = IntegerArithmetic.primeDecomp(n, primes);
		int result = 1;
		for (int i = 0; i < primeDec.size(); i++) {
			result *= (primeDec.get(i).get(0) - 1) * (int)Math.pow(primeDec.get(i).get(0), primeDec.get(i).get(1) - 1);
		}
		return result;
	}
}

File: problem72/problem72.js
/**
 * Problem 72
 */

var IntegerArithmetic = require('../libraries/IntegerArithmetic.js');
var upperBound = Math.floor(Math.sqrt(1000000));
var prime = IntegerArithmetic.sieveEratosthenes(upperBound);
var nbPrimes = 0;
for (var i = 0; i < prime.length; i++) {
	if (prime[i]) nbPrimes++;
}
var primes = [];
var j = 0;
for (var i = 0; i < prime.length; i++) {
	if (prime[i]) primes.push(i);
}	
var sum = 0;
for (var d = 2; d <= 1000000; d++) {
	sum += phi(d, primes);
}
console.log('The solution is ' + sum);

/**
* Returns phi n
* @param integer n
* @param array of integers primes
* @return integer
*/
function phi(n, primes) {
	var primeDec = IntegerArithmetic.primeDecomp(n, primes);
	var result = 1;
	for (var i = 0; i < primeDec.length; i++) {
		result *= (primeDec[i][0] - 1) * Math.floor(Math.pow(primeDec[i][0], primeDec[i][1] - 1));
	}
	return result;
}